qi-lib/on.rkt

```diff
@@ -22,13 +22,14 @@
    #`((flow clause) #,@(syntax->list #'ags))])
 
 (define-syntax-parser flow-lambda
-  [(_ (arg:id ...) clause:clause)
-   #'(lambda (arg ...)
-       (on (arg ...)
-           clause))]
   [(_ rest-args:id clause:clause)
    #'(lambda rest-args
        (on (rest-args)
+           clause))]
+  [(_ args:formals clause:clause)
+   #:with ags (attribute args.params)
+   #'(lambda args
+       (on ags
```

Comment:
When I first saw `☯`, I thought about whether it's needed to add some keywords to it (such as `#:name`, `#:args`, `#:keywords`, `#:kw-formals`), then I saw `flow-lambda` and I guess `qi` prefers to use `flow-lambda` to define functions (I can't think of any other reason to use `flow-lambda` instead of `☯` at the moment). And I do prefer to use `☯` directly (I love the symbol!), so instead of avoiding the double-wrapping, can we try to add more keywords to `☯`?

Reply:
How would adding keywords to `☯` help, could you give an example?

Reply:
For example:
```racket
(define t
  (☯ #:name 't
     #:formals (#:a a [b 0])
     (~>
      (ε (~>> 1> unbox add1 (set-box! a)))
      cons)))

(define t
  (π (#:a a [b 0])
    (~>
     (ε (~>> 1> unbox add1 (set-box! a)))
     cons)))

(t #:a (box 1))     ; '(#&2 . 0)
(t #:a (box 1) 2)   ; '(#&2 . 2)
(object-name t)     ; 't
(procedure-arity t) ; '(0 1)
```

Reply:
And I think we can make `☯`'s rest-args (declared via `#:formals` ) to be `values`, `flow-lambda` to stay the same.

Reply:
My personal feeling is that those keywords give some friction with the spirit of `flow`, but I do agree that there's an interesting question here of the difference between `flow` and `flow-lambda`. The distinction seems to be that `flow-lambda` permits bindings via parameters.

I'm not sure you can avoid double-wrapping immediately: you need a binding form (`lambda`) and a flow. The `on` macro is exactly an immediately applied flow. Giving `flow` the ability to bind names might avoid the double-wrapping. In a hypothetical syntax:
```racket
;; Instead of
(lambda args (on ags flo))
;; which is equivalent to
(lambda args ((flow flo) ags))
;; we can write a flow with bindings
((flow #:formals args flo) ags)
```
But it still feels like a conflation of concerns. The `flow` macro doesn't bind parameters (well, publicly; the implementation has to) because that's what it's supposed to do (?).

Also: the naïve translation of the last example is the same as the first two examples. Is there a more efficient expansion? Perhaps depending on `flo`, but that requires more analysis, no?

Reply:
> you need a binding form (lambda) and a flow. The on macro is exactly an immediately applied flow.

I think you are right, if we want new binding forms we have to use `λ`. I guess there is probably no way to git rid of double-wrapping (in a sense enabling flow to bind names is just another form of double-wrapping).

> My personal feeling is that those keywords give some friction with the spirit of flow

I'm new to `qi` so maybe I'm wrong: In my opinion, `☯` is used to translate a flo-expr to a procedure, so that it makes sense to me to add some keywords to customize the returned procedure.

Following is some translations in my mind:

```racket
(☯ #:arity arity #:name name #:realm realm flo) => (procedure-reduce-arity (☯ flo) arity name realm)
(☯ #:defun (name args) flo) => (letrec ([name (λ args (on ags (~>> (apply values) flo)))]) name)
```

Reply:
I'll let @countvajhula respond with more detail, this being their project, but in my mind Qi is less about translating flows and more about expressing a certain kind of computation (namely, computations over values). That this happens to be done via translation to procedures is incidental, to me.

Reply:
In general Qi isn't too concerned with being a DSL for Racket functions -- more about functions and values abstractly. So, I'm doubtful we'd want to include anything too specific to Racket's implementation as part of the language. I'm working on a doc as part of the upcoming Qi release that may shed some light on what Qi is trying to be, and I think we're all learning more about it as we go 🙂 

At the same time, notions like arity will be useful internally for optimization purposes because Racket's compiler can optimize better when it knows the arity, so it's not unlikely that an arity-aware dialect of Qi would be a good intermediate representation in the compiler.

Thanks for confirming that there's no way to eliminate the double-wrapping! When Qi gets native bindings (#36 ) we should have more options here.